// caudex/dist/index.umd.js
// caudex v0.0.6 - https://github.com/wikibonsai/caudex.git
import { withTimeout, Mutex } from 'async-mutex';
import { customAlphabet, nanoid } from 'nanoid';

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }
  return obj;
}

// caudex information //

// data structures
let DATA_STRUCT;

// node information ('node kind', 'node type')

// eslint-disable-next-line @typescript-eslint/no-namespace
(function (DATA_STRUCT) {
  DATA_STRUCT["BASE"] = "base";
  DATA_STRUCT["TREE"] = "tree";
  DATA_STRUCT["WEB"] = "web";
})(DATA_STRUCT || (DATA_STRUCT = {}));
let NODE;

// relationship information (rel kind) //

// eslint-disable-next-line @typescript-eslint/no-namespace
(function (_NODE) {
  let KIND;
  (function (KIND) {
    KIND["DOC"] = "doc";
    KIND["MEDIA"] = "media";
    KIND["TEMPLATE"] = "template";
    KIND["ZOMBIE"] = "zombie";
  })(KIND || (KIND = {}));
  _NODE.KIND = KIND;
  let TYPE;
  (function (TYPE) {
    TYPE["DEFAULT"] = "default";
    TYPE["ENTRY"] = "entry";
    TYPE["INDEX"] = "index";
  })(TYPE || (TYPE = {}));
  _NODE.TYPE = TYPE;
  let MEDIA;
  (function (MEDIA) {
    MEDIA["MARKDOWN"] = "markdown";
  })(MEDIA || (MEDIA = {}));
  _NODE.MEDIA = MEDIA;
  let STATE;
  (function (STATE) {
    STATE["ORPHAN"] = "orphan";
    STATE["ISOLATE"] = "isolate";
  })(STATE || (STATE = {}));
  _NODE.STATE = STATE;
})(NODE || (NODE = {}));
let REL;
(function (_REL) {
  let FAM;
  (function (FAM) {
    FAM["FAM"] = "fam";
    FAM["ANCESTORS"] = "ancestors";
    FAM["PARENT"] = "parent";
    FAM["SIBLINGS"] = "siblings";
    FAM["CHILDREN"] = "children";
    FAM["DESCENDANTS"] = "descendants";
    FAM["LINEAGE"] = "lineage";
  })(FAM || (FAM = {}));
  _REL.FAM = FAM;
  let REF;
  (function (REF) {
    REF["REF"] = "ref";
    REF["ATTR"] = "attr";
    REF["LINK"] = "link";
    REF["EMBED"] = "embed";
  })(REF || (REF = {}));
  _REL.REF = REF;
  let LEVEL;
  (function (LEVEL) {
    LEVEL["FILE"] = "file";
    LEVEL["HEADER"] = "header";
  })(LEVEL || (LEVEL = {}));
  _REL.LEVEL = LEVEL;
})(REL || (REL = {}));
let QUERY_TYPE;
(function (QUERY_TYPE) {
  QUERY_TYPE["ID"] = "id";
  QUERY_TYPE["NODE"] = "node";
  QUERY_TYPE["NODEKIND"] = "nodekind";
  QUERY_TYPE["NODETYPE"] = "nodetype";
  QUERY_TYPE["DATA"] = "data";
  QUERY_TYPE["ZOMBIE"] = "zombie";
})(QUERY_TYPE || (QUERY_TYPE = {}));

class Node {
  // relationships

  constructor(id, kind, type, data) {
    this.id = id;
    this.kind = kind;
    this.type = type;
    this.data = data;
    // relationships
    // tree
    this.children = [];
    // web
    this.attrs = {};
    this.links = [];
    this.embeds = [];
  }
  flush(kind) {
    if (!kind || kind === DATA_STRUCT.TREE) {
      this.children = [];
    }
    if (!kind || kind === DATA_STRUCT.WEB) {
      this.attrs = {};
      this.links = [];
      this.embeds = [];
    }
  }

  // note: ('in-' methods are used in the 'web' for generating 'attributed' and 'backlinks')

  inAttrs(id) {
    return Object.values(this.attrs).filter(ids => ids.has(id)).length !== 0;
  }
  inChildren(id) {
    return this.children.includes(id);
  }
  inLinks(id, header) {
    return this.links.filter(link => link.id === id && (header === undefined || link.header === header)).length !== 0;
  }
  inEmbeds(id, header) {
    return this.embeds.filter(embed => embed.id === id && (header === undefined || embed.header === header)).length !== 0;
  }
}

class Base {
  // since there are no pointers...pseudo-pointers:
  //  reference:   node properties -- 'childID'
  //  dereference: class methods   -- 'this.get(childID)'

  // note: mixins don't work with protected/private properties...
  // the core index instance
  // node data properties that are unique get a hash value for faster access
  // key opts
  // node data key that should be unique

  // nanoid options

  constructor(items, opts) {
    _defineProperty(this, "uniqKeys", []);
    _defineProperty(this, "zombieKey", '');
    // go
    // options
    if (opts) {
      // unique node data keys
      if (!opts.uniqKeys) {
        console.warn('no "uniqKeys" given, this may effect access speeds in some cases');
        this.uniqKeyMap = undefined;
      } else {
        this.uniqKeys = opts.uniqKeys;
        this.uniqKeyMap = {};
        for (const key of this.uniqKeys) {
          this.uniqKeyMap[key] = {};
        }
      }
      // todo: force 'zombieKey' to be a 'uniqKeys'?...'query' needs it...
      // zombie handling
      if (!opts.zombieKey) {
        console.warn('no "zombieKey" given, this will result in zombie nodes with empty data keys and may negatively effect cases where nodes referencing the zombie node are expecting that data key to exist');
      } else {
        this.zombieKey = opts.zombieKey;
      }
      // nanoid
      if (opts.nanoid) {
        if (!opts.nanoid.alphabet || !opts.nanoid.size) {
          throw new Error('when using "nanoid" options, please be sure to fill in both "alphabet" and "size" properties');
        } else {
          this.nanoidOpts = {
            alphabet: opts.nanoid.alphabet,
            size: opts.nanoid.size
          };
        }
      }
    }
    // init items; populate nodes
    this.index = {};
    const errorItems = [];
    for (const item of items) {
      const newNode = this.add(item.data, item.init);
      if (newNode === undefined) {
        errorItems.push(item);
      }
    }
    // initialize mutex
    if (opts && opts.thread) {
      this.useLock = opts.thread.safe ? opts.thread.safe : false;
      // @ts-expect-error: 'Property '_semaphore' is missing in type 'MutexInterface' but required in type 'Mutex'.ts(2741)'
      this.lock = opts.thread.timeout ? withTimeout(new Mutex(), opts.thread.timeout) : new Mutex();
    } else {
      this.useLock = false;
      this.lock = new Mutex(); // setting the mutex is just to make types happy...
    }
    // finally, print data that was not successfully initialized
    if (errorItems.length > 0) {
      throw new Error(`unable to create nodes from items:\n${JSON.stringify(errorItems)}`);
    }
  }

  // util

  checkLock() {
    if (this.useLock && !this.lock.isLocked()) {
      throw new Error('please lock the index to access it');
    }
  }
  genID() {
    const configdNanoid = this.nanoidOpts ? customAlphabet(this.nanoidOpts.alphabet, this.nanoidOpts.size) : nanoid;
    return configdNanoid();
  }
  print() {
    let printout = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    this.checkLock();
    if (printout) {
      console.log(JSON.stringify(this.index));
    }
    return JSON.stringify(this.index);
  }

  // properties

  applyFilter(nodes, filter) {
    if (!filter) {
      return nodes;
    }
    return nodes.filter(node => {
      var _node$data;
      if (filter.nodeKind !== undefined && node.kind !== filter.nodeKind) {
        return false;
      }
      if (filter.nodeType !== undefined && node.type !== filter.nodeType) {
        return false;
      }
      if (filter.filename !== undefined && ((_node$data = node.data) === null || _node$data === void 0 ? void 0 : _node$data.filename) !== filter.filename) {
        return false;
      }
      return true;
    });
  }
  resolvePayload(id, payload, node) {
    var _n$data3;
    const n = node !== null && node !== void 0 ? node : this.index[id];
    if (!n && (payload === QUERY_TYPE.DATA || payload === QUERY_TYPE.NODEKIND || payload === QUERY_TYPE.NODETYPE || typeof payload === 'string' && payload !== QUERY_TYPE.ID)) {
      return undefined;
    }
    if (payload === undefined || payload === QUERY_TYPE.ID) {
      return id;
    }
    if (payload === QUERY_TYPE.NODE) {
      return n;
    }
    if (payload === QUERY_TYPE.NODEKIND) {
      return n === null || n === void 0 ? void 0 : n.kind;
    }
    if (payload === QUERY_TYPE.NODETYPE) {
      return n === null || n === void 0 ? void 0 : n.type;
    }
    if (payload === QUERY_TYPE.DATA) {
      return n === null || n === void 0 ? void 0 : n.data;
    }
    if (payload === QUERY_TYPE.ZOMBIE) {
      var _n$data;
      return n === null || n === void 0 ? void 0 : (_n$data = n.data) === null || _n$data === void 0 ? void 0 : _n$data[this.zombieKey];
    }
    if (Array.isArray(payload)) {
      const res = {};
      for (const key of payload) {
        var _n$data2;
        res[key] = key in ((_n$data2 = n === null || n === void 0 ? void 0 : n.data) !== null && _n$data2 !== void 0 ? _n$data2 : {}) ? n.data[key] : this.execQuery(id, key);
      }
      return res;
    }
    if (Object.keys((_n$data3 = n === null || n === void 0 ? void 0 : n.data) !== null && _n$data3 !== void 0 ? _n$data3 : {}).includes(payload)) {
      return n.data[payload];
    }
    return this.execQuery(id, payload);
  }
  all(opts) {
    var _opts$payload;
    this.checkLock();
    const payload = (_opts$payload = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload !== void 0 ? _opts$payload : QUERY_TYPE.ID;
    let nodes = Object.values(this.index);
    nodes = this.applyFilter(nodes, opts === null || opts === void 0 ? void 0 : opts.filter);
    return nodes.map(node => this.resolvePayload(node.id, payload, node));
  }
  nodetypes() {
    var _ref;
    this.checkLock();
    /* eslint-disable indent */
    const nodes = (_ref = this.all({
      payload: QUERY_TYPE.NODE
    })) !== null && _ref !== void 0 ? _ref : [];
    const nodetypes = nodes.filter(node => node.type !== undefined && node.type !== '').map(node => node.type);
    /* eslint-enable indent */
    return new Set(nodetypes);
  }
  zombies(opts) {
    this.checkLock();
    const mergedFilter = {
      ...(opts === null || opts === void 0 ? void 0 : opts.filter),
      nodeKind: NODE.KIND.ZOMBIE
    };
    return this.all({
      ...opts,
      filter: mergedFilter
    });
  }

  // index operations

  has(id) {
    this.checkLock();
    return Object.keys(this.index).includes(id);
  }
  flushData(id) {
    this.checkLock();
    // single
    if (id) {
      const node = this.get(id, {
        payload: QUERY_TYPE.NODE
      });
      if (node === undefined) {
        return false;
      }
      node.data = {};
      return true;
      // all
    } else {
      var _ref2;
      ((_ref2 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref2 !== void 0 ? _ref2 : []).forEach(node => {
        node.data = {};
      });
      return true;
    }
  }
  flushRels() {
    this.checkLock();
    for (const node of (_ref3 = this.all({
      payload: QUERY_TYPE.NODE
    })) !== null && _ref3 !== void 0 ? _ref3 : []) {
      var _ref3;
      // delete zombies
      if (node.kind === NODE.KIND.ZOMBIE) {
        delete this.index[node.id];
        // flush
      } else {
        node.flush();
      }
    }
    return true;
  }
  clear() {
    this.checkLock();
    this.index = {};
  }

  // node operations

  // validate

  // validate unique data properties are unique
  // add "id" when validating node data to be updated.
  validate(data) {
    let id = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
    this.checkLock();
    for (const key of Object.keys(data)) {
      if (this.uniqKeys.includes(key)) {
        if (Object.values(this.index).find(node => node.data[key] === data[key] && node.id !== id)) {
          console.warn(`data key "${key}" with value "${data[key]}" already exists`);
          return false;
        }
      }
    }
    return true;
  }

  // add

  // zombie-case
  add(data, init) {
    this.checkLock();
    // does id exist?
    if (data.id && this.has(data.id)) {
      console.warn(`node with id "${data.id}" already exists`);
      return undefined;
    }
    // zombie-case
    if (typeof data === 'string') {
      const zombieData = {};
      zombieData[this.zombieKey] = data;
      const zombieNode = new Node(this.genID(), NODE.KIND.ZOMBIE, undefined, zombieData);
      this.index[zombieNode.id] = zombieNode;
      if (this.uniqKeyMap) {
        this.uniqKeyMap[this.zombieKey][data] = zombieNode.id;
      }
      return zombieNode;
    }
    // default-case
    // is data valid?
    if (!this.validate(data)) {
      // warning will print in 'validateNodeData()' call
      return undefined;
    }
    const id = init && init.id ? init.id : this.genID();
    const kind = init && init.kind ? init.kind : NODE.KIND.DOC;
    const type = init && init.type ? init.type : NODE.TYPE.DEFAULT;
    // init
    const newNode = new Node(id, kind, type, data);
    this.index[newNode.id] = newNode;
    // populate 'uniqKeyMap'
    for (const key of Object.keys(data)) {
      if (this.uniqKeyMap && Object.keys(this.uniqKeyMap).includes(key)) {
        this.uniqKeyMap[key][data[key]] = id;
      }
    }
    return newNode;
  }

  // edit

  edit(id, key, newValue) {
    this.checkLock();
    const node = this.index[id];
    if (node === undefined) {
      console.warn(`node with id "${id}" does not exist`);
      return false;
    }
    if (key === QUERY_TYPE.NODETYPE) {
      node.type = newValue;
      if (node.type === newValue) {
        return true;
      }
    } else {
      const data = node.data;
      if (this.uniqKeyMap && this.uniqKeys.includes(key)) {
        delete this.uniqKeyMap[key][data[key]];
        this.uniqKeyMap[key][newValue] = id;
      }
      data[key] = newValue;
      if (!this.validate(data, id)) {
        return false;
      }
      node.data[key] = newValue;
      if (node.data[key] === newValue) {
        return true;
      }
    }
    return false;
  }
  fill(id, data) {
    this.checkLock();
    if (!this.has(id) && !(data.id && this.has(data.id))) {
      console.warn(`node with id "${id}" does not exist`);
      return undefined;
    }
    if (!this.validate(data, id)) {
      return undefined;
    }
    for (const key of Object.keys(data)) {
      if (this.uniqKeyMap && Object.keys(this.uniqKeyMap).includes(key)) {
        this.uniqKeyMap[key][data[key]] = id;
      }
    }
    this.index[id].data = data;
    this.index[id].kind = data.kind ? data.kind : NODE.KIND.DOC;
    this.index[id].type = data.type ? data.type : NODE.TYPE.DEFAULT;
    return this.index[id];
  }

  // get

  get(id, opts) {
    var _opts$payload2;
    this.checkLock();
    if (!this.index[id]) {
      console.warn(`node with id "${id}" does not exist`);
      return undefined;
    }
    const payload = (_opts$payload2 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload2 !== void 0 ? _opts$payload2 : QUERY_TYPE.NODE;
    return this.resolvePayload(id, payload, this.index[id]);
  }
  execQuery(id, qType) {
    this.checkLock();
    if (qType === QUERY_TYPE.ID) {
      return id;
    } else if (qType === QUERY_TYPE.NODE) {
      return this.index[id];
    } else if (qType === QUERY_TYPE.NODEKIND) {
      return this.index[id].kind;
    } else if (qType === QUERY_TYPE.NODETYPE) {
      return this.index[id].type;
      // todo: make it possible to have a query type for dynamically defined data keys
    } else if (qType === QUERY_TYPE.DATA) {
      return this.index[id].data;
    } else if (qType === QUERY_TYPE.ZOMBIE) {
      return this.index[id].data[this.zombieKey];
    } else if (Object.keys(this.index[id].data).includes(qType)) {
      return this.index[id].data[qType];
    } else {
      if (!this.index[id].data[this.zombieKey]) {
        console.warn(`query failed for id "${id}" with type '${qType}'`);
      }
      // todo: ...?
      // if (this.index[id].data.zombie) {
      //   console.debug(`query encountered an unexpected zombie with id: "${id}"`);
      // } else {
      //   console.warn(`query failed for id "${id}" with type '${type}'`);
      // }
    }
  }

  find(key, value) {
    this.checkLock();
    if (!this.uniqKeys.includes(key)) {
      console.warn('"dataKey" must be a unique key in node data. Did you mean "filter()"?');
      return undefined;
    }
    if (this.uniqKeyMap && Object.keys(this.uniqKeyMap).includes(key)) {
      const id = this.uniqKeyMap[key][value];
      return this.index[id];
    }
    return Object.values(this.index).find(node => {
      return node.data[key] && node.data[key] === value;
    });
  }
  filter(key, value) {
    this.checkLock();
    if (key === QUERY_TYPE.NODEKIND) {
      return Object.values(this.index).filter(node => {
        return node.kind === value;
      });
    } else if (key === QUERY_TYPE.NODETYPE) {
      return Object.values(this.index).filter(node => {
        return node.type === value;
      });
    } else {
      return Object.values(this.index).filter(node => {
        return node.data[key] && node.data[key] === value;
      });
    }
  }

  // rm

  rm(id) {
    var _ref4;
    this.checkLock();
    const node = this.get(id, {
      payload: QUERY_TYPE.NODE
    });
    if (!node) {
      console.warn(`node with id "${id}" does not exist`);
      return false;
    }
    const hasRel = ((_ref4 = this.all({
      payload: QUERY_TYPE.NODE
    })) !== null && _ref4 !== void 0 ? _ref4 : []).some(n => n.id !== id && (n.inChildren(id) || n.inAttrs(id) || n.inLinks(id)));
    // if other nodes reference this node, just delete data
    if (!hasRel) {
      for (const key of Object.keys(node.data)) {
        if (this.uniqKeyMap && Object.keys(this.uniqKeyMap).includes(key)) {
          delete this.uniqKeyMap[key][node.data[key]];
        }
      }
      delete this.index[id];
      if (!this.has(id)) {
        return true;
      }
      // todo: when we are just about to cleanup relationships, 'hasRel' will be true, but only for nodes that are about to be deleted.
    } else {
      var _this$get;
      for (const key of Object.keys(node.data)) {
        if (this.uniqKeyMap && Object.keys(this.uniqKeyMap).includes(key) && key !== this.zombieKey) {
          delete this.uniqKeyMap[key][node.data[key]];
        }
      }
      const zombieData = node.data[this.zombieKey];
      node.kind = NODE.KIND.ZOMBIE;
      node.type = undefined;
      node.data = {
        [this.zombieKey]: zombieData
      };
      if (node.kind === NODE.KIND.ZOMBIE && (_this$get = this.get(id)) !== null && _this$get !== void 0 && _this$get.data[this.zombieKey]) {
        return true;
      }
    }
    return false;
  }
}

function Web(Base) {
  return class Web extends Base {
    // properties

    // nodes that are not linked to any other node in the web
    isolates(opts) {
      var _opts$payload, _ref;
      this.checkLock();
      const payload = (_opts$payload = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload !== void 0 ? _opts$payload : QUERY_TYPE.ID;
      /* eslint-disable indent */
      return ((_ref = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref !== void 0 ? _ref : []).filter(node => {
        var _this$neighbors;
        return ((_this$neighbors = this.neighbors(node.id, opts)) === null || _this$neighbors === void 0 ? void 0 : _this$neighbors.length) === 0 && node.kind !== NODE.KIND.ZOMBIE;
      }).map(node => this.get(node.id, {
        ...opts,
        payload
      }));
      /* eslint-enable indent */
    }

    // web-lvl

    reftypes() {
      var _ref2;
      this.checkLock();
      let reftypes = [];
      /* eslint-disable indent */
      ((_ref2 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref2 !== void 0 ? _ref2 : []).forEach(node => {
        const attrTypes = Object.keys(node.attrs).map(type => type);
        // @ts-expect-error: typescript is not smart enough to see 'filter' performing validation
        const linkTypes = node.links.filter(link => link !== undefined).map(link => link.type);
        reftypes = reftypes.concat(attrTypes).concat(linkTypes);
      });
      /* eslint-enable indent */
      return new Set(reftypes);
    }
    attrtypes() {
      var _ref3;
      this.checkLock();
      let attrtypes = [];
      /* eslint-disable indent */
      ((_ref3 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref3 !== void 0 ? _ref3 : []).forEach(node => attrtypes = attrtypes.concat(Object.keys(node.attrs).map(type => type)));
      /* eslint-enable indent */
      return new Set(attrtypes);
    }
    linktypes() {
      var _ref4;
      this.checkLock();
      let linktypes = [];
      /* eslint-disable indent */
      ((_ref4 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref4 !== void 0 ? _ref4 : []).forEach(node =>
      // @ts-expect-error: typescript is not smart enough to see 'filter' performing validation
      linktypes = linktypes.concat(node.links.filter(link => link !== undefined).map(link => link.type)));
      /* eslint-enable indent */
      return new Set(linktypes);
    }

    // node-lvl

    // // types
    // forerefs(id: string, query?: QueryType.id): Attrs | undefined;
    // forerefs(id: string, query: QueryType.node): Record<string, Node[]> | undefined;
    // forerefs(id: string, query: string | string[]): Record<string, any> | undefined;
    // // define
    // forerefs(id: string, query: string | string[] = QueryType.id): Attrs | Record<string, Node[]> | Record<string, any> | undefined {
    // }

    // backrefs(id: string, query?: QueryType.id): Attrs | undefined;
    // backrefs(id: string, query: QueryType.node): Record<string, Node[]> | undefined;
    // backrefs(id: string, query: string | string[]): Record<string, any> | undefined;
    // // define
    // backrefs(id: string, query: string | string[] = QueryType.id): Attrs | Record<string, Node[]> | Record<string, any> | undefined {
    // }

    foreattrs(id, opts) {
      var _opts$filter, _opts$filter2, _opts$payload2;
      this.checkLock();
      if ((opts === null || opts === void 0 ? void 0 : (_opts$filter = opts.filter) === null || _opts$filter === void 0 ? void 0 : _opts$filter.header) !== undefined || (opts === null || opts === void 0 ? void 0 : (_opts$filter2 = opts.filter) === null || _opts$filter2 === void 0 ? void 0 : _opts$filter2.level) === REL.LEVEL.HEADER) {
        throw new Error('attrs do not support headers');
      }
      const node = this.get(id, {
        payload: QUERY_TYPE.NODE
      });
      if (!node) {
        return undefined;
      }
      const payload = (_opts$payload2 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload2 !== void 0 ? _opts$payload2 : QUERY_TYPE.ID;
      if (payload === QUERY_TYPE.ID || payload === undefined) {
        return node.attrs;
      }
      const queryPayload = {};
      for (const [type, ids] of Object.entries(node.attrs)) {
        queryPayload[type] = [];
        for (const targetId of ids) {
          queryPayload[type].push(this.get(targetId, {
            ...opts,
            payload
          }));
        }
      }
      return queryPayload;
    }
    backattrs(id, opts) {
      var _opts$filter3, _opts$filter4, _opts$payload3;
      this.checkLock();
      if ((opts === null || opts === void 0 ? void 0 : (_opts$filter3 = opts.filter) === null || _opts$filter3 === void 0 ? void 0 : _opts$filter3.header) !== undefined || (opts === null || opts === void 0 ? void 0 : (_opts$filter4 = opts.filter) === null || _opts$filter4 === void 0 ? void 0 : _opts$filter4.level) === REL.LEVEL.HEADER) {
        throw new Error('attrs do not support headers');
      }
      if (!this.has(id)) {
        return undefined;
      }
      const backattrs = {};
      for (const node of (_ref5 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref5 !== void 0 ? _ref5 : []) {
        var _ref5;
        if (node.inAttrs(id)) {
          for (const [type, ids] of Object.entries(node.attrs)) {
            if (ids.has(id)) {
              if (!backattrs[type]) {
                backattrs[type] = new Set();
              }
              backattrs[type].add(node.id);
            }
          }
        }
      }
      const payload = (_opts$payload3 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload3 !== void 0 ? _opts$payload3 : QUERY_TYPE.ID;
      if (payload === QUERY_TYPE.ID || payload === undefined) {
        return backattrs;
      }
      const queryPayload = {};
      for (const [type, ids] of Object.entries(backattrs)) {
        queryPayload[type] = [];
        for (const sourceId of ids) {
          queryPayload[type].push(this.get(sourceId, {
            ...opts,
            payload
          }));
        }
      }
      return queryPayload;
    }
    forelinks(id, opts) {
      var _opts$payload4;
      this.checkLock();
      const node = this.get(id, {
        payload: QUERY_TYPE.NODE
      });
      if (!node) {
        return undefined;
      }
      const f = opts === null || opts === void 0 ? void 0 : opts.filter;
      let links = node.links;
      if ((f === null || f === void 0 ? void 0 : f.header) !== undefined) {
        links = links.filter(l => l.header === f.header);
      }
      if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.FILE) {
        links = links.filter(l => !l.header);
      }
      if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.HEADER) {
        links = links.filter(l => !!l.header);
      }
      if ((f === null || f === void 0 ? void 0 : f.type) !== undefined) {
        links = links.filter(l => l.type === f.type);
      }
      const payload = (_opts$payload4 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload4 !== void 0 ? _opts$payload4 : QUERY_TYPE.ID;
      if (payload === QUERY_TYPE.ID || payload === undefined) {
        return links;
      }
      return links.map(link => [link.type, this.get(link.id, {
        ...opts,
        payload
      })]);
    }
    backlinks(id, opts) {
      var _opts$payload5;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const backlinks = [];
      for (const node of (_ref6 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref6 !== void 0 ? _ref6 : []) {
        var _ref6;
        if (node.inLinks(id)) {
          for (const link of node.links) {
            if (id !== link.id) {
              continue;
            }
            const f = opts === null || opts === void 0 ? void 0 : opts.filter;
            if ((f === null || f === void 0 ? void 0 : f.header) !== undefined && link.header !== f.header) {
              continue;
            }
            if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.FILE && link.header) {
              continue;
            }
            if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.HEADER && !link.header) {
              continue;
            }
            if ((f === null || f === void 0 ? void 0 : f.type) !== undefined && link.type !== f.type) {
              continue;
            }
            backlinks.push({
              type: link.type,
              id: node.id,
              ...(link.header !== undefined && {
                header: link.header
              })
            });
          }
        }
      }
      const payload = (_opts$payload5 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload5 !== void 0 ? _opts$payload5 : QUERY_TYPE.ID;
      if (payload === QUERY_TYPE.ID || payload === undefined) {
        return backlinks;
      }
      return backlinks.map(link => [link.type, this.get(link.id, {
        ...opts,
        payload
      })]);
    }
    foreembeds(id, opts) {
      var _opts$payload6;
      this.checkLock();
      const node = this.get(id, {
        payload: QUERY_TYPE.NODE
      });
      if (!node) {
        return undefined;
      }
      const f = opts === null || opts === void 0 ? void 0 : opts.filter;
      let embeds = node.embeds;
      if ((f === null || f === void 0 ? void 0 : f.header) !== undefined) {
        embeds = embeds.filter(e => e.header === f.header);
      }
      if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.FILE) {
        embeds = embeds.filter(e => !e.header);
      }
      if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.HEADER) {
        embeds = embeds.filter(e => !!e.header);
      }
      const payload = (_opts$payload6 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload6 !== void 0 ? _opts$payload6 : QUERY_TYPE.ID;
      if (payload === QUERY_TYPE.ID || payload === undefined) {
        return embeds;
      }
      return embeds.map(embed => this.get(embed.id, {
        ...opts,
        payload
      }));
    }
    backembeds(id, opts) {
      var _opts$payload7;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const backembeds = [];
      for (const node of (_ref7 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref7 !== void 0 ? _ref7 : []) {
        var _ref7;
        if (node.inEmbeds(id)) {
          for (const embed of node.embeds) {
            var _embed$media;
            if (id !== embed.id) {
              continue;
            }
            const f = opts === null || opts === void 0 ? void 0 : opts.filter;
            if ((f === null || f === void 0 ? void 0 : f.header) !== undefined && embed.header !== f.header) {
              continue;
            }
            if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.FILE && embed.header) {
              continue;
            }
            if ((f === null || f === void 0 ? void 0 : f.level) === REL.LEVEL.HEADER && !embed.header) {
              continue;
            }
            backembeds.push({
              id: node.id,
              media: (_embed$media = embed.media) !== null && _embed$media !== void 0 ? _embed$media : NODE.MEDIA.MARKDOWN,
              ...(embed.header !== undefined && {
                header: embed.header
              })
            });
          }
        }
      }
      const payload = (_opts$payload7 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload7 !== void 0 ? _opts$payload7 : QUERY_TYPE.ID;
      if (payload === QUERY_TYPE.ID || payload === undefined) {
        return backembeds;
      }
      return backembeds.map(embed => this.get(embed.id, {
        ...opts,
        payload
      }));
    }
    neighbors(id, kindOrOpts) {
      var _filter$kind, _filter;
      this.checkLock();
      const kind = typeof kindOrOpts === 'string' && Object.values(REL.REF).includes(kindOrOpts) ? kindOrOpts : (_filter$kind = kindOrOpts === null || kindOrOpts === void 0 ? void 0 : (_filter = kindOrOpts.filter) === null || _filter === void 0 ? void 0 : _filter.kind) !== null && _filter$kind !== void 0 ? _filter$kind : REL.REF.REF;
      let neighbors = [];
      const node = this.get(id, {
        payload: QUERY_TYPE.NODE
      });
      if (!node) {
        return undefined;
      }
      if (kind === REL.REF.REF || kind === REL.REF.ATTR) {
        var _this$backattrs;
        const backattrs = (_this$backattrs = this.backattrs(id)) !== null && _this$backattrs !== void 0 ? _this$backattrs : {};
        neighbors = neighbors.concat(Object.values(node.attrs).flatMap(ids => Array.from(ids))).concat(Object.values(backattrs).flatMap(ids => Array.from(ids)));
      }
      if (kind === REL.REF.REF || kind === REL.REF.LINK) {
        var _this$backlinks;
        const backlinks = (_this$backlinks = this.backlinks(id)) !== null && _this$backlinks !== void 0 ? _this$backlinks : [];
        neighbors = neighbors.concat(node.links.map(link => link.id)).concat(backlinks.map(link => link.id));
      }
      if (kind === REL.REF.REF || kind === REL.REF.EMBED) {
        var _this$backembeds;
        const backembeds = (_this$backembeds = this.backembeds(id)) !== null && _this$backembeds !== void 0 ? _this$backembeds : [];
        neighbors = neighbors.concat(node.embeds.map(embed => embed.id)).concat(backembeds.map(embed => embed.id));
      }
      return neighbors;
    }

    // methods

    flushRelRefs(id) {
      this.checkLock();
      // single
      if (id) {
        const node = this.get(id);
        if (node === undefined) {
          return false;
        }
        // delete zombies...
        /* eslint-disable indent */
        const relRefIDs = Array.from(Object.values(node.attrs).flatMap(ids => Array.from(ids)).concat(node.links.map(link => link.id).concat(node.embeds.map(embed => embed.id))));
        for (const relRefID of relRefIDs) {
          var _ref8;
          const relRefNode = this.get(relRefID, {
            payload: QUERY_TYPE.NODE
          });
          if (!relRefNode) {
            continue;
          }
          const isZombie = relRefNode.kind === NODE.KIND.ZOMBIE;
          const hasRel = ((_ref8 = this.all({
            payload: QUERY_TYPE.NODE
          })) !== null && _ref8 !== void 0 ? _ref8 : []).some(n => n.id !== relRefID && n.inChildren(relRefID) || n.id !== relRefID && n.id !== id && (n.inAttrs(relRefID) || n.inLinks(relRefID) || n.inEmbeds(relRefID)));
          /* eslint-enable indent */
          // delete floater/orphaned zombies
          if (isZombie && !hasRel) {
            delete this.index[relRefID];
          }
          // flush
          node.flush(DATA_STRUCT.WEB);
        }
        return true;
        // all
      } else {
        for (const node of (_ref9 = this.all({
          payload: QUERY_TYPE.NODE
        })) !== null && _ref9 !== void 0 ? _ref9 : []) {
          var _ref9, _ref10;
          const isZombie = node.kind === NODE.KIND.ZOMBIE;
          const hasFamRel = ((_ref10 = this.all({
            payload: QUERY_TYPE.NODE
          })) !== null && _ref10 !== void 0 ? _ref10 : []).some(relNode => relNode.id !== node.id && relNode.inChildren(node.id));
          // delete floater/orphaned zombies
          if (isZombie && !hasFamRel) {
            delete this.index[node.id];
            return true;
          }
          // flush
          node.flush(DATA_STRUCT.WEB);
        }
        return true;
      }
    }

    // add

    connect(sourceID, targetID, optsOrKind, typeOrMedia) {
      var _ref11;
      const opts = typeof optsOrKind === 'string' && Object.values(REL.REF).includes(optsOrKind) ? {
        kind: optsOrKind,
        type: optsOrKind === REL.REF.EMBED ? undefined : typeOrMedia !== null && typeOrMedia !== void 0 ? typeOrMedia : '',
        media: optsOrKind === REL.REF.EMBED ? (_ref11 = typeOrMedia) !== null && _ref11 !== void 0 ? _ref11 : NODE.MEDIA.MARKDOWN : undefined
      } : optsOrKind;
      const {
        kind,
        type = '',
        header,
        media
      } = opts;
      if (kind === REL.REF.REF) {
        console.warn('please connect to a more specific relationship(\'REL.REF.ATTR\', \'REL.REF.LINK\', or \'REL.REF.EMBED\'');
        return false;
      }
      if (kind === REL.REF.ATTR && header !== undefined) {
        throw new Error('attrs do not support headers');
      }
      this.checkLock();
      const sourceNode = this.get(sourceID, {
        payload: QUERY_TYPE.NODE
      });
      if (!sourceNode) {
        console.warn(`source node with id "${sourceID}" not found`);
        return false;
      }
      if (!this.has(targetID)) {
        console.warn(`target node with id "${targetID}" not found`);
        return false;
      }
      if (kind === REL.REF.ATTR) {
        if (!Object.keys(sourceNode.attrs).includes(type)) {
          sourceNode.attrs[type] = new Set([targetID]);
        } else {
          sourceNode.attrs[type].add(targetID);
        }
        return sourceNode.attrs[type].has(targetID);
      }
      if (kind === REL.REF.LINK) {
        const hasLink = sourceNode.links.find(link => link.type === type && link.id === targetID && link.header === header);
        if (hasLink === undefined) {
          sourceNode.links.push({
            type,
            id: targetID,
            ...(header !== undefined && {
              header
            })
          });
        }
        return true;
      }
      if (kind === REL.REF.EMBED) {
        const resolvedMedia = media !== null && media !== void 0 ? media : NODE.MEDIA.MARKDOWN;
        if (resolvedMedia !== NODE.MEDIA.MARKDOWN && !Object.values(NODE.MEDIA).includes(resolvedMedia)) {
          console.warn('invalid media kind: ' + resolvedMedia);
          return false;
        }
        const hasEmbed = sourceNode.embeds.find(embed => embed.media === resolvedMedia && embed.id === targetID && embed.header === header);
        if (hasEmbed === undefined) {
          sourceNode.embeds.push({
            media: resolvedMedia,
            id: targetID,
            ...(header !== undefined && {
              header
            })
          });
        }
        return true;
      }
      return false;
    }

    // edit

    disconnect(sourceID, targetID, optsOrKind, typeOrMedia) {
      var _ref12;
      const opts = typeof optsOrKind === 'string' && Object.values(REL.REF).includes(optsOrKind) ? {
        kind: optsOrKind,
        type: optsOrKind === REL.REF.EMBED ? undefined : typeOrMedia !== null && typeOrMedia !== void 0 ? typeOrMedia : '',
        media: optsOrKind === REL.REF.EMBED ? (_ref12 = typeOrMedia) !== null && _ref12 !== void 0 ? _ref12 : NODE.MEDIA.MARKDOWN : undefined
      } : optsOrKind;
      const {
        kind,
        type = '',
        header,
        media
      } = opts;
      if (kind === REL.REF.REF) {
        console.warn('please disconnect a more specific relationship(\'REL.REF.ATTR\', \'REL.REF.LINK\', or \'REL.REF.EMBED\'');
        return false;
      }
      this.checkLock();
      const sourceNode = this.get(sourceID, {
        payload: QUERY_TYPE.NODE
      });
      if (!sourceNode) {
        console.warn(`source node with id "${sourceID}" not found`);
        return false;
      }
      if (!this.has(targetID)) {
        console.warn(`target node with id "${targetID}" not found`);
        return false;
      }
      if (kind === REL.REF.ATTR) {
        var _sourceNode$attrs$typ, _sourceNode$attrs$typ2, _sourceNode$attrs$typ3;
        if (((_sourceNode$attrs$typ = sourceNode.attrs[type]) === null || _sourceNode$attrs$typ === void 0 ? void 0 : _sourceNode$attrs$typ.size) === 1) {
          delete sourceNode.attrs[type];
          return !Object.keys(sourceNode.attrs).includes(type);
        }
        (_sourceNode$attrs$typ2 = sourceNode.attrs[type]) === null || _sourceNode$attrs$typ2 === void 0 ? void 0 : _sourceNode$attrs$typ2.delete(targetID);
        return !((_sourceNode$attrs$typ3 = sourceNode.attrs[type]) !== null && _sourceNode$attrs$typ3 !== void 0 && _sourceNode$attrs$typ3.has(targetID));
      }
      if (kind === REL.REF.LINK) {
        for (let i = 0; i < sourceNode.links.length; i++) {
          const l = sourceNode.links[i];
          if (l.id === targetID && l.type === type && l.header === header) {
            sourceNode.links.splice(i, 1);
            return true;
          }
        }
        return !sourceNode.links.find(link => link.type === type && link.id === targetID && link.header === header);
      }
      if (kind === REL.REF.EMBED) {
        const resolvedMedia = media !== null && media !== void 0 ? media : NODE.MEDIA.MARKDOWN;
        for (let i = 0; i < sourceNode.embeds.length; i++) {
          const e = sourceNode.embeds[i];
          if (e.id === targetID && e.media === resolvedMedia && e.header === header) {
            sourceNode.embeds.splice(i, 1);
            return true;
          }
        }
        return !sourceNode.embeds.find(embed => embed.media === (media !== null && media !== void 0 ? media : NODE.MEDIA.MARKDOWN) && embed.id === targetID && embed.header === header);
      }
      return false;
    }
    retype(oldType, newType) {
      let kind = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : REL.REF.REF;
      this.checkLock();
      const retypes = [];
      for (const node of (_ref13 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref13 !== void 0 ? _ref13 : []) {
        var _ref13;
        if (kind === REL.REF.REF || kind === REL.REF.ATTR) {
          if (Object.keys(node.attrs).includes(oldType) && !Object.keys(node.attrs).includes(newType)) {
            node.attrs[newType] = new Set(node.attrs[oldType]);
            delete node.attrs[oldType];
            retypes.push(Object.keys(node.attrs).includes(newType));
          }
        }
        if (kind === REL.REF.REF || kind === REL.REF.LINK) {
          for (const l of node.links) {
            if (l.type === oldType) {
              l.type = newType;
              retypes.push(l.type === newType);
            }
          }
        }
      }
      return retypes.every(r => r === true);
    }
    transfer(sourceID, targetID) {
      let kind = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : REL.REF.REF;
      this.checkLock();
      if (sourceID === targetID) {
        console.warn('source and target are the same');
        return false;
      }
      const sourceNode = this.get(sourceID, {
        payload: QUERY_TYPE.NODE
      });
      const targetNode = this.get(targetID, {
        payload: QUERY_TYPE.NODE
      });
      if (!sourceNode) {
        console.warn(`source node with id "${sourceID}" not in index`);
        return false;
      }
      if (!targetNode) {
        console.warn(`target node with id "${targetID}" not in index`);
        return false;
      }
      if (kind === REL.REF.REF || kind === REL.REF.ATTR) {
        for (const [key, val] of Object.entries(sourceNode.attrs)) {
          // create
          if (!Object.keys(targetNode.attrs).includes(key)) {
            targetNode.attrs[key] = val;
            // add
          } else {
            val.forEach(v => {
              if (v !== targetID) {
                targetNode.attrs[key].add(v);
              }
            });
          }
        }
        sourceNode.attrs = {};
      }
      if (kind === REL.REF.REF || kind === REL.REF.LINK) {
        targetNode.links = targetNode.links.concat(sourceNode.links.filter(l => l.id !== targetID));
        sourceNode.links = [];
      }
      if (kind === REL.REF.REF || kind === REL.REF.EMBED) {
        targetNode.embeds = targetNode.embeds.concat(sourceNode.embeds.filter(e => e.id !== targetID));
        sourceNode.embeds = [];
      }
      return true;
    }
  };
}

function Tree(Base) {
  return class Tree extends Base {
    // root operations

    // note: typescript does not allow for separate types between get/set
    //  - https://github.com/microsoft/TypeScript/issues/43662
    //  - https://github.com/microsoft/TypeScript/issues/32821
    //  - https://github.com/microsoft/TypeScript/issues/2521

    setRoot(id) {
      this.checkLock();
      if (!this.has(id)) {
        console.warn(`node with id "${id}" not in index`);
        return false;
      }
      this._root = id;
      if (this._root) {
        return true;
      }
      return false;
    }

    // todo...?
    // public hasRoot(): boolean {
    //   const hasRoot: boolean = (this._root !== undefined);
    //   if (!hasRoot) { console.warn("root undefined"); }
    //   return hasRoot;
    // }

    // properties

    // tree-lvl

    root(opts) {
      this.checkLock();
      if (this._root === undefined) {
        return undefined;
      }
      // default payload for root() is ID (unlike get() which defaults to NODE)
      return this.get(this._root, opts !== undefined ? opts : {
        payload: QUERY_TYPE.ID
      });
    }
    orphans(treeIDs, opts) {
      var _opts$payload, _ref;
      this.checkLock();
      const payload = (_opts$payload = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload !== void 0 ? _opts$payload : QUERY_TYPE.ID;
      /* eslint-disable indent */
      return ((_ref = this.all({
        ...opts,
        payload: QUERY_TYPE.NODE
      })) !== null && _ref !== void 0 ? _ref : []).filter(node => treeIDs.includes(node.id) && node.children.length === 0 && !this.parent(node.id) && node.kind !== NODE.KIND.ZOMBIE).map(node => this.get(node.id, {
        ...opts,
        payload
      }));
      /* eslint-enable indent */
    }

    ancestors(id, opts) {
      var _opts$payload2;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ids = this.getRelFam(id, REL.FAM.ANCESTORS);
      const payload = (_opts$payload2 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload2 !== void 0 ? _opts$payload2 : QUERY_TYPE.ID;
      return payload === QUERY_TYPE.ID || payload === undefined ? ids : ids.map(nodeId => this.get(nodeId, {
        ...opts,
        payload
      }));
    }
    parent(id, opts) {
      var _opts$payload3;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ids = this.getRelFam(id, REL.FAM.PARENT);
      if (ids.length === 0) {
        return '';
      }
      const payload = (_opts$payload3 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload3 !== void 0 ? _opts$payload3 : QUERY_TYPE.ID;
      return payload === QUERY_TYPE.ID || payload === undefined ? ids[0] : this.get(ids[0], {
        ...opts,
        payload
      });
    }
    siblings(id, opts) {
      var _opts$payload4;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ids = this.getRelFam(id, REL.FAM.SIBLINGS);
      const payload = (_opts$payload4 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload4 !== void 0 ? _opts$payload4 : QUERY_TYPE.ID;
      return payload === QUERY_TYPE.ID || payload === undefined ? ids : ids.map(nodeId => this.get(nodeId, {
        ...opts,
        payload
      }));
    }
    children(id, opts) {
      var _opts$payload5;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ids = this.getRelFam(id, REL.FAM.CHILDREN);
      const payload = (_opts$payload5 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload5 !== void 0 ? _opts$payload5 : QUERY_TYPE.ID;
      return payload === QUERY_TYPE.ID || payload === undefined ? ids : ids.map(nodeId => this.get(nodeId, {
        ...opts,
        payload
      }));
    }
    descendants(id, opts) {
      var _opts$payload6;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ids = this.getRelFam(id, REL.FAM.DESCENDANTS);
      const payload = (_opts$payload6 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload6 !== void 0 ? _opts$payload6 : QUERY_TYPE.ID;
      return payload === QUERY_TYPE.ID || payload === undefined ? ids : ids.map(nodeId => this.get(nodeId, {
        ...opts,
        payload
      }));
    }
    lineage(id, opts) {
      var _opts$payload7;
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ids = this.getRelFam(id, REL.FAM.LINEAGE);
      const payload = (_opts$payload7 = opts === null || opts === void 0 ? void 0 : opts.payload) !== null && _opts$payload7 !== void 0 ? _opts$payload7 : QUERY_TYPE.ID;
      return payload === QUERY_TYPE.ID || payload === undefined ? ids : ids.map(nodeId => this.get(nodeId, {
        ...opts,
        payload
      }));
    }
    level(id) {
      this.checkLock();
      if (!this.has(id)) {
        return undefined;
      }
      const ancestors = this.ancestors(id, {
        payload: QUERY_TYPE.ID
      });
      if (ancestors === undefined) {
        return undefined;
      }
      return ancestors.length;
    }

    // util

    getRelFam(id, relFam) {
      this.checkLock();
      // if uri does not exist in index
      if (!this.has(id)) {
        return [];
      }
      // prep data collection
      const initRelFamData = {};
      initRelFamData[relFam] = [];
      // get target family uris
      const famIDs = this.search(id, initRelFamData);
      return famIDs === undefined ? [] : famIDs[relFam];
    }

    // todo
    // public getFamRels(id: string, relFams: REL.FAM[]): Record<REL.FAM, string[]> {
    //   // if uri does not exist in index
    //   if (!this.has(id)) { return {} as Record<REL.FAM, string[]>; }
    //   // prep data collection
    //   const initFamRelsData: Record<REL.FAM, string[]> = {} as Record<REL.FAM, string[]>;
    //   for (const famRel of relFams) {
    //     initFamRelsData[famRel] = [];
    //   }
    //   // get target family uris
    //   const relFamsData: Record<REL.FAM, string[]> | undefined = this.search(id, initFamRelsData);
    //   return (relFamsData === undefined) ? {} as Record<REL.FAM, string[]> : relFamsData;
    // }

    // usage note: use getter wrappers plz
    search(nodeID, relData) {
      let node = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      let depth = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
      let found = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : false;
      this.checkLock();
      // handle root + init result info
      if (depth === 0) {
        node = this.root({
          payload: QUERY_TYPE.NODE
        });
        if (node === undefined) {
          return undefined;
        }
      }
      const atTargetNode = nodeID === node.id;
      if (atTargetNode || found) {
        // at target node
        if (atTargetNode) {
          if (REL.FAM.CHILDREN in relData) {
            relData['children'] = node.children;
          }
        }
        // already found -- continue building 'descendants' / 'lineage'
        if (REL.FAM.DESCENDANTS in relData) {
          relData['descendants'] = relData['descendants'].concat(node.children);
        }
        if (REL.FAM.LINEAGE in relData) {
          relData['lineage'] = relData['lineage'].concat(node.children);
        }
        if (REL.FAM.DESCENDANTS in relData || REL.FAM.LINEAGE in relData) {
          for (const child of node.children) {
            const nextNode = this.get(child, {
              payload: QUERY_TYPE.NODE
            });
            if (nextNode === undefined) {
              return undefined;
            }
            this.search(nodeID, relData, nextNode, depth + 1, true);
          }
        }
        return relData;
        // still searching or building
      } else {
        if (REL.FAM.ANCESTORS in relData) {
          relData['ancestors'].push(node.id);
        }
        if (REL.FAM.LINEAGE in relData) {
          relData['lineage'].push(node.id);
        }
        // if current node is the target's parent
        const targetNodeID = node.children.find(child => child === nodeID);
        if (targetNodeID !== undefined) {
          const targetNode = this.get(targetNodeID, {
            payload: QUERY_TYPE.NODE
          });
          if (targetNode === undefined) {
            return undefined;
          }
          if (REL.FAM.PARENT in relData) {
            relData['parent'] = [node.id];
          }
          if (REL.FAM.SIBLINGS in relData) {
            relData['siblings'] = node.children.filter(child => child !== nodeID);
          }
          const result = this.search(nodeID, relData, targetNode, depth + 1);
          if (result !== undefined) {
            return result;
          }
          // keep searching
        } else {
          for (const child of node.children) {
            const nextNode = this.get(child, {
              payload: QUERY_TYPE.NODE
            });
            if (nextNode === undefined) {
              return undefined;
            }
            // there should only be one unique path, so return only that one result
            const result = this.search(nodeID, JSON.parse(JSON.stringify(relData)), nextNode, depth + 1);
            if (result !== undefined) {
              return result;
            }
          }
        }
      }
    }

    // methods

    flushRelFams() {
      this.checkLock();
      for (const node of (_ref2 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref2 !== void 0 ? _ref2 : []) {
        var _ref2, _ref3;
        const isZombie = node.kind === NODE.KIND.ZOMBIE;
        /* eslint-disable indent */
        const hasRelRef = ((_ref3 = this.all({
          payload: QUERY_TYPE.NODE
        })) !== null && _ref3 !== void 0 ? _ref3 : []).some(relNode => relNode.id !== node.id && (relNode.inAttrs(node.id) || relNode.inLinks(node.id)));
        /* eslint-enable indent */
        // delete orphaned zombies
        if (isZombie && !hasRelRef) {
          delete this.index[node.id];
          return true;
        }
        // flush
        node.flush(DATA_STRUCT.TREE);
      }
      return true;
    }

    // add

    // (this method is for all the asian pears out there 🍐)
    graft(parentID, childID) {
      let force = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      this.checkLock();
      if (parentID === childID) {
        console.warn('parentID and childID are the same');
        return false;
      }
      if (!this.has(parentID)) {
        console.warn(`parent node with id "${parentID}" not in index`);
        return false;
      }
      if (!this.has(childID)) {
        console.warn(`child node with id "${childID}" not in index`);
        return false;
      }
      if (this.inTree(childID)) {
        console.warn(`child node with id "${childID}" already exists in the tree`);
        return false;
      }
      this.get(parentID, {
        payload: QUERY_TYPE.NODE
      }).children.push(childID);
      if (!force && !this.isTree()) {
        this.get(parentID, {
          payload: QUERY_TYPE.NODE
        }).children.pop();
        return false;
      }
      return true;
    }

    // edit

    // replace the 'source' with the 'target', so in the end the 'target' is in the tree.
    replace(sourceID, targetID) {
      this.checkLock();
      if (sourceID === targetID) {
        console.warn('source and target are the same');
        return false;
      }
      const sourceNode = this.get(sourceID, {
        payload: QUERY_TYPE.NODE
      });
      const targetNode = this.get(targetID, {
        payload: QUERY_TYPE.NODE
      });
      if (!sourceNode) {
        console.warn(`source node with id "${sourceID}" not in index`);
        return false;
      }
      if (!targetNode) {
        console.warn(`target node with id "${targetID}" not in index`);
        return false;
      }
      // target already in tree
      /* tsignore */
      if (this.parent(targetID) !== '' || targetNode.children.length !== 0) {
        console.warn(`target with "id" "${targetID}" already exists in tree`);
        return false;
      }
      // parent
      const parentNode = this.parent(sourceID, {
        payload: QUERY_TYPE.NODE
      });
      if (!parentNode) {
        console.warn(`no parent exists for ${JSON.stringify(sourceNode)}`);
        return false;
      } else {
        const sourceIndex = parentNode.children.indexOf(sourceID);
        parentNode.children[sourceIndex] = targetID;
      }
      // children
      targetNode.children = sourceNode.children;
      sourceNode.children = [];
      return true;
    }

    // good for handling single index doc edits
    // 'subrootID' should be included in the 'subtree' param
    // note: does not handle zombie creation
    transplant(subrootID, subtree) {
      if (!subtree.find(node => node.id === subrootID)) {
        console.warn(`subroot with id "${subrootID}" not found in the subtree`);
        return false;
      }
      this.checkLock();
      const subrootNode = this.get(subrootID, {
        payload: QUERY_TYPE.NODE
      });
      if (!subrootNode) {
        console.warn(`subroot with id "${subrootID}" not found in the index`);
        return false;
      }
      const rollbackState = new Map();
      const newSubtreeMap = new Map(subtree.map(node => [node.id, node.children]));
      // perform changes
      this.doTransplant(subrootID, newSubtreeMap, rollbackState);
      // rollback if invalid tree or some other error
      if (!this.isTree()) {
        console.warn('transplant failed due to invalid resultant tree -- rolling back to previous state');
        this.rollback(rollbackState);
        return false;
      }
      return true;
    }
    doTransplant(nodeID, newSubtreeMap, rollbackState) {
      const node = this.get(nodeID, {
        payload: QUERY_TYPE.NODE
      });
      if (!node) {
        return;
      }
      // save state for possible rollback
      rollbackState.set(nodeID, {
        children: [...node.children]
      });
      const newChildIDs = newSubtreeMap.get(nodeID);
      if (newChildIDs !== undefined) {
        // update
        node.children = newChildIDs;
        this.index[nodeID] = node;
        // recurse
        node.children.forEach(childID => this.doTransplant(childID, newSubtreeMap, rollbackState));
      }
    }

    // rollback for subtree transplants
    rollback(rollbackState) {
      for (const [nodeID, originalState] of rollbackState) {
        const node = this.get(nodeID, {
          payload: QUERY_TYPE.NODE
        });
        if (node === undefined) {
          console.warn(`node with id "${nodeID}" not found in the index when performing rollback`);
          continue;
        }
        node.children = originalState.children;
        this.index[nodeID] = node;
      }
    }

    // rm

    prune(parentID, childID) {
      let force = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      this.checkLock();
      if (parentID === childID) {
        console.warn('parentID and childID are the same');
        return false;
      }
      if (!this.has(parentID)) {
        console.warn(`parent node with id "${parentID}" not in index`);
        return false;
      }
      if (!this.has(childID)) {
        console.warn(`child node with id "${childID}" not in index`);
        return false;
      }
      // check if the node to be pruned is the root or has children
      if (!force && (this.isRoot(childID) || !this.isLeaf(childID))) {
        console.warn('cannot prune root or non-leaf child node');
        return false;
      }
      const parentNode = this.get(parentID, {
        payload: QUERY_TYPE.NODE
      });
      const childIndex = parentNode.children.indexOf(childID);
      if (childIndex === -1) {
        console.warn(`child node with id "${childID}" is not a child of parent "${parentID}"`);
        return false;
      }
      // Remove the child from the parent's children array
      parentNode.children.splice(childIndex, 1);
      // If the resulting structure is not a valid tree, revert the change
      if (!force && !this.isTree()) {
        parentNode.children.push(childID);
        return false;
      }
      return true;
    }

    // tree utils

    inTree(id) {
      var _ref4;
      this.checkLock();
      /* eslint-disable indent */
      return ((_ref4 = this.all({
        payload: QUERY_TYPE.NODE
      })) !== null && _ref4 !== void 0 ? _ref4 : []).find(node => node.children.includes(id)) !== undefined;
      /* eslint-enable indent */
    }

    isRoot(id) {
      this.checkLock();
      return id === this.root();
    }
    isLeaf(id) {
      var _this$children;
      this.checkLock();
      return ((_this$children = this.children(id)) === null || _this$children === void 0 ? void 0 : _this$children.length) === 0;
    }
    isTree() {
      let curNode = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.root({
        payload: QUERY_TYPE.NODE
      });
      let visited = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : new Set();
      this.checkLock();
      if (curNode === undefined) {
        console.warn('node is undefined');
        return false;
      }
      if (visited.has(curNode.id)) {
        console.warn(`node with id "${curNode.id}" already visited`);
        return false;
      }
      visited.add(curNode.id);
      for (const childId of curNode.children) {
        const childNode = this.get(childId, {
          payload: QUERY_TYPE.NODE
        });
        if (childNode === undefined) {
          console.warn(`node with id "${childId}" not found`);
          return false;
        }
        if (!this.isTree(childNode, visited)) {
          return false;
        }
      }
      return true;
    }

    // 'key' -- the data key to print
    printTree(key) {
      let printout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
      this.checkLock();
      const rootNode = this.root({
        payload: QUERY_TYPE.NODE
      });
      if (rootNode === undefined) {
        throw new Error('root undefined');
      }
      const treeStr = this.buildTreeString(key, rootNode, '', true);
      if (printout) {
        console.log(treeStr);
      }
      return treeStr;
    }
    buildTreeString(key, node) {
      let prefix = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '';
      let isRoot = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
      let result = isRoot ? `${node.id}: ${JSON.stringify(node.data[key]) || 'node not found'}\n` : '';
      node.children.forEach((childID, index) => {
        const childNode = this.get(childID, {
          payload: QUERY_TYPE.NODE
        });
        if (childNode === undefined) {
          return;
        }
        const isLastChild = index === node.children.length - 1;
        const childPrefix = prefix + (isLastChild ? '└── ' : '├── ');
        const grandchildPrefix = prefix + (isLastChild ? '    ' : '│   ');
        const subtree = this.buildTreeString(key, childNode, grandchildPrefix);
        result += childPrefix + `${childNode.id}: ${JSON.stringify(childNode.data[key]) || 'node not found'}\n` + subtree;
      });
      return result;
    }
  };
}

// export as 'static property mixin'
// from: https://www.typescriptlang.org/docs/handbook/mixins.html#static-property-mixins-17829
class Caudex extends Web(Tree(Base)) {}

export { Base, Caudex, DATA_STRUCT, NODE, Node, QUERY_TYPE, REL, Tree, Web };
